### PR TITLE
Java sample app

### PR DESCRIPTION
**Pull request by evanli02**
Description:

Link to tracking Issue:

Testing Performed:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by bryan-aguilar** [comment]
Please update your `.gitignore` file to ignore IDEA files and also Gradle build files. You can find a set of `gitignore` templates [here](https://github.com/github/gitignore)

File: settings.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+include(":dependencyManagement")
+include(":instrumentation:logback-1.0")
+include(":instrumentation:log4j-2.13.2")
+include(":otelagent")
+include(":smoke-tests:fakebackend")
+include(":smoke-tests:runner")
+include(":smoke-tests:spring-boot")
+include(":sample-apps:springboot")
+include(":sample-apps:auto-instrumentation-java-spark")
+include(":sample-apps:spark-awssdkv1")
```
Comment (bryan-aguilar) [inline settings.gradle.kts:51]: Some of these include directives are not valid anymore. There are no `smoke-tests` anymore. I believe these are remnants from the copy.

**Comment by bryan-aguilar** [comment]
It looks like the github workflows reference some directories that are not a part of this repository. Also, are the github workflow secrets valid in this repository?

File: .github/workflows/main-build.yml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+          registry: public.ecr.aws
+
+      - name: Run test containers
+        run: docker-compose up --abort-on-container-exit
+        working-directory: .github/collector
+        env:
+          INSTANCE_ID: ${{ github.run_id }}-${{ github.run_number }}
+          LISTEN_ADDRESS: 0.0.0.0:4567
+          APP_IMAGE: public.ecr.aws/aws-otel-test/aws-otel-java-spark:${{ github.sha }}
+          VALIDATOR_COMMAND: -c spark-otel-trace-metric-validation.yml --endpoint http://app:4567 --metric-namespace aws-otel-integ-test -t ${{ github.run_id }}-${{ github.run_number }}
```
Comment (bryan-aguilar) [inline .github/workflows/main-build.yml:114]: This validation will fail with the changes to the output metrics.

**Comment by bryan-aguilar** [comment]
Please add the `.gradle` directory to the gitignore and remove those tracked files. See [here](https://github.com/github/gitignore/blob/main/Gradle.gitignore) for an example gradle `.gitignore` file.

**Review by rapphil — COMMENTED** [review]
Do we really need to upload the files in the .idea directory? Could that entire directory be placed in the `.gitignore`?

**Review by rapphil — COMMENTED** [review]
1. I think only the sample app directory should be a gradle project as this repo will contain sample apps for different languages as well, each with its own build system.

2. Please make sure to add to gitignore and to not commit the following directories in any of the subtrees:
* `.gradle`
* `.idea`

3. Add a readme file with the minimal steps to have the application up and running from the command line.

4. Can you use spotless in your project to auto format the code and add it permanently as a development dependency? https://github.com/diffplug/spotless/tree/main/plugin-gradle

File: sample-apps/auto-instrumentation-java-spark/src/main/resources/config.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+Host: "127.0.0.1"                       # Host - String Address
+port: "4567"                          # Port - String Port
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/resources/config.yaml:3]:
```suggestion
Port: "4567"                          # Port - String Port
```

This should be capitalized

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/Config.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    private static int configThreads;
+    private static int configCpu;
+    private static ArrayList<String> configSamplePorts;
+    public Config() {
+        //reading in values from config yaml file
+        Yaml yaml = new Yaml();
+        InputStream inputStream = getClass().getResourceAsStream("/config.yaml");
+        Map<String, Object> variables = yaml.load(inputStream);
+        configHost = (String) variables.get("Host");
+        configPort = (String) variables.get("port");
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/Config.java:27]:
```suggestion
        configPort = (String) variables.get("Port");
```
With the previous suggestion, this should also be updated

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.FileNotFoundException;
+
+public class MetricEmitter {
+  private static final Logger logger = LogManager.getLogger();
+  private static Config getConfig() { return new Config(); }
+  static final AttributeKey<String> DIMENSION_API_NAME = AttributeKey.stringKey("apiName");
+  static final AttributeKey<String> DIMENSION_STATUS_CODE = AttributeKey.stringKey("statusCode");
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:18]: These metric attribute names are outdated. The new ones are following this key-value convention 
{
“signal”:     (string)“metric”
“language”:   (string)<name of language used>
“metricType”: (string)“random” or (string)"request"
}

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        totalHeapSizeName = API_RANDOM_ASYNC_UPDOWN_METRIC + "_" + instanceId;
+        threadsActiveName = API_RANDOM_UPDOWN_METRIC + "_" + instanceId;
+        cpuUsageName = API_RANDOM_GAUGE + "_" + instanceId;
+    }
+
+    //building synchronous request-based counter metric
+    bytesCounter = meter
+            .counterBuilder(totalBytesSentName)
+            .setDescription("API request load sent in bytes")
+            .setUnit("mb")
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:82]:
```suggestion
            .setDescription("Keeps a sum of the total amount of bytes sent while the application is alive")
            .setUnit("By")
```

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            .build();
+
+    //building asynchronous request-based counter metric
+    meter
+            .counterBuilder(totalApiRequestsName)
+            .setDescription("Total amount of API requests to endpoint")
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:88]:
```suggestion
            .setDescription("Increments by one every time a sampleapp endpoint is used")
```

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+            .setUnit("1")
+            .buildWithCallback(measurement ->
+                    measurement.record(
+                            totalApiRequests,
+                            Attributes.of(DIMENSION_API_NAME, apiNameValue, DIMENSION_STATUS_CODE, statusCodeValue)));
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:93]: replace these attributes with the ones mentioned, following the requirements doc standard

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+    //building histogram request-based metric
+    apiLatencyHistogram = meter
+            .histogramBuilder(latencyTimeName)
+            .setDescription("API latency time")
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:98]:
```suggestion
            .setDescription("Measures latency time in buckets of 100 300 and 500”)
```

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+            .setUnit("ms")
+            .build();
+
+    //building synchronous random-based counter metric
+    timeCounter = meter
+            .counterBuilder(timeAliveName)
+            .setDescription("How Long Application is Alive")
+            .setUnit("s")
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:106]:
```suggestion
            .setDescription("Total amount of time that the application has been alive")
            .setUnit("ms")
```
This also requires to change the method that updates `timeAlive` to reflect "ms". 
By convention, I believe we can only use "By" "ms" and "1" as units

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+            .build();
+
+    //building asynchronous random-based up-down counter metric
+    meter
+            .upDownCounterBuilder(totalHeapSizeName)
+            .setDescription("Heap size")
+            .setUnit("1")
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:113]:
```suggestion
            .setDescription("The current total heap size")
            .setUnit("By")
```

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+            .buildWithCallback(measurement ->
+                    measurement.record(
+                            totalHeapSize,
+                            Attributes.of(DIMENSION_API_NAME, apiNameValue, DIMENSION_STATUS_CODE, statusCodeValue)));
+
+    //building synchronous random-based up-down counter metric
+    threadsCounter = meter
+            .counterBuilder(threadsActiveName)
+            .setDescription("Number of Threads Active")
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:122]:
```suggestion
            .setDescription("The total number of threads active")
```

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+            .setUnit("1")
+            .build();
+
+    //building random-based gauge metric
+    meter
+            .gaugeBuilder(cpuUsageName)
+            .setDescription("Measures CPU Usage")
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:129]:
```suggestion
            .setDescription("Cpu usage percent")
```

File: sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+            .setUnit("1")
+            .ofLongs()
+            .buildWithCallback(measurement ->
+                    measurement.record(
+                            cpuUsage,
+                            Attributes.of(DIMENSION_API_NAME, apiNameValue, DIMENSION_STATUS_CODE, statusCodeValue)));
```
Comment (alexperez52) [inline sample-apps/auto-instrumentation-java-spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:135]: These attributes also have to be updated

**Comment by alexperez52** [comment]
Left some suggestions that align with the requirements doc

**Comment by rapphil** [comment]
Hi @evanli02 are you going to address the comments in the CR?

If not I will continue your work.

**Comment by evanli02** [comment]
Hi Raphael,

I'm currently busy with classes and other commitments so I, unfortunately,
do not think I will have time to address the comments. Sorry about that.

Evan

**Comment by rapphil** [comment]
Hi @evanli02, thanks for replying. I will take over and continue the work. Cheers!

**Pull request closed without merging**